chalk-solve/src/coherence.rs

```diff
@@ -106,18 +106,20 @@ where
 
     // Build the forest of specialization relationships.
     fn build_specialization_forest(&self) -> Result<Graph<ImplId<I>, ()>, CoherenceError<I>> {
-        // The forest is returned as a graph but built as a GraphMap; this is
-        // so that we never add multiple nodes with the same ItemId.
-        let mut forest = DiGraphMap::new();
+        let mut forest = DiGraph::new();
 
         // Find all specializations (implemented in coherence/solve)
         // Record them in the forest by adding an edge from the less special
         // to the more special.
         self.visit_specializations_of_trait(|less_special, more_special| {
-            forest.add_edge(less_special, more_special, ());
+            // Check so that we never add multiple nodes with the same ImplId.
+            let l = forest.add_node(less_special);
```

Comment:
If we didn't add the nodes before, why do we need to add them now? update_edge doesn't do anything different versus add_edge with respect to nodes.

Reply:
The nodes must be added because [`petgraph::graphmap::GraphMap::add_edge`](https://docs.rs/petgraph/latest/petgraph/graphmap/struct.GraphMap.html#method.add_edge) can take a generic type `N` (that must implement `Ord`), where [`petgraph::graph::Graph::update_edge`](https://docs.rs/petgraph/latest/petgraph/graph/struct.Graph.html#method.update_edge) takes a [`NodeIndex`](https://docs.rs/petgraph/latest/petgraph/graph/struct.NodeIndex.html).

Reply:
Ah, I was looking at Graph::add_edge.

Can this create duplicate nodes?

Reply:
It's not entirely clear from the docs whether Graph::add_node will create "duplicate" nodes... https://docs.rs/petgraph/latest/petgraph/graph/struct.Graph.html#method.add_node

Reply:
Graph::add_node does create multiple nodes given the same value:

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=fe37070d0747054726917b2ba2cd2e96

Reply:
I'm going to see if I can handle this logic with a match statement: 

```
let node_impls: Vec<ImplId<_>> = forest.raw_nodes().iter().map(|x| x.weight).collect();
self.visit_specializations_of_trait(|less_special, more_special| {
    match (
        node_impls.contains(&less_special),
        node_impls.contains(&more_special),
    ) {
        (true, true) => todo!(),
        (true, false) => todo!(),
        (false, true) => todo!(),
        (false, false) => todo!(),
    }
}
```

Reply:
BTW, [Graph::update_edge](https://docs.rs/petgraph/0.6.0/petgraph/graph/struct.Graph.html#method.update_edge) is only used to update the data for an edge. In this case, the edges are all `()`, so this doesn't help us.

Reply:
I see "Add or update an edge from a to b." - So it's basically `add_edge` without allowing duplicates.

Reply:
Seems like it might make sense to keep around a `Map<ImplId, NodeIndex>`

Reply:
Before `visit_specialization_of_trait`, `forest` will be empty...

Reply:
I'm trying to reason this out. The logic below seems alright, but I'm not sure how to get the indices for existing nodes...

```
// Find all specializations. (This is implemented in coherence/solve.)
// Record them in the forest by adding an edge from the less special
// to the more special.
self.visit_specializations_of_trait(|less_special, more_special| {
    match (
        node_impls.contains(&less_special),
        node_impls.contains(&more_special),
    ) {
        (true, true) => {
            // but how do we get indices for l and m?
            forest.add_edge(l, m, ());
        }
        (true, false) => {
            let m = forest.add_node(more_special);
            // but how do we get index for l?
            forest.add_edge(l, m, ());
        }
        (false, true) => {
            let l = forest.add_node(less_special);
            // but how do we get index for m?
            forest.add_edge(l, m, ());
        }
        (false, false) => {
            // add L and M
            let l = forest.add_node(less_special);
            let m = forest.add_node(more_special);
            forest.add_edge(l, m, ());
        }
    }
})?;
```

Reply:
Can this even be done correctly without GraphMap? 😬

Reply:
I'm thinking something like
```rust
let mut forest = DiGraph::new();
let mut node_map = FxHashMap::new();
self.visit_specializations_of_trait(|less_special, more_special| {
    let less_special_node = *node_map.entry(less_special).or_insert_with(|| forest.add_node(less_special));
    let more_special_node = *node_map.entry(more_special).or_insert_with(|| forest.add_node(more_special));
    forest.update_edge(less_special_noe, more_special_node, ());
}
```